_blogposts/2023-05-17-enhancing-the-ergonomics-of-records.mdx

```diff
@@ -0,0 +1,140 @@
+
+However, there's always room for improvement. In ReScript v10 we added support for [optional record fields](https://rescript-lang.org/docs/manual/latest/record#optional-record-fields) to make constructing records with many optional fields easier and more ergonomic. Now we're focusing on how we can improve the workflow around defining records.
+
+In ReScript v11, we're delighted to introduce two new features that will significantly boost your record manipulation capabilities: Record Type Spreads and Record Type Coercion. Let's delve into what these new features offer and how they can make working with records more ergonomic.
+
+## Type Spreads
+
+Before type spreads, creating a new record that was similar or extended from another record required spelling out every single field in the new definition. This was often tedious, error-prone and made code harder to maintain, especially when working with records with many fields.
+
+In ReScript v11, you can now spread one or more record types into a new record type. It looks like this:
+
+```rescript
+type a = {
+  id: string,
+  name: string,
+}
+
+type b = {
+  age: int
+}
+
+type c = {
+  ...a,
+  ...b,
+  active: bool
+}
+```
+
+`type c` will now be:
+
+```rescript
+type c = {
+  id: string,
+  name: string,
+  age: int,
+  active: bool,
+}
+```
+
+Keeping it as straightforward as possible, spreads are essentially a 'copy-paste' operation for fields from one or more records to another, inlining the fields from the spread records into the new record.
+
+This is going to be a much more ergonomic experience when working with types with many fields, where variations of the same underlying type are needed.
+
+### Use case: extending the built in DOM nodes
+
+This feature can be particularly useful when extending DOM nodes. For instance, in the case of the animation library Framer Motion, one could easily extend the native DOM types with additional properties specific to the library, leading to a more seamless and type-safe integration.
+
+This is how you could bind to a `div` in Framer Motion with the new record type spreads:
+
+```rescript
+type animate = {} // definition omitted for brevity
+
+type divProps = {
+  ...JsxDOM.domProps,
+  initial?: animate,
+  animate?: animate,
+  whileHover?: animate,
+  whileTap?: animate,
+}
+
+module Div = {
+  @module("framer-motion") external make: divProps => Jsx.element = "div"
+}
+```
+
+You can now use `<Div />` as a `<motion.div />` component from Framer Motion. And your type definition is now simple and easy to maintain.
+
+## Type Coercion
+
+Type coercion introduces an extra layer of flexibility when working with records. Now, records of the same shape can be coerced between each other. This means that we can cast record `a` to be record `b` at the type level, if they contain the same exact fields. An example:
```

Comment:
Maybe worth noting that type coercion has already been a thing for polymorphic variants so far. In ReScript 11, we extended _type-safe_ coercion to records as well.

(special note on type-safe, since TypeScript only does unsafe type coercion with the `as` operator)

Reply:
Also I realized we never covered type-coercion for structural objects in the docs...

https://rescript-lang.org/try?version=v10.1.2&code=C4TwDgpgBAhlC8UDeAoKUBEA7GBbCGAXFAM7ABOAllgOYA0amMNBx1wKAviiqJFACMEyRtjytSFajShceAGwjAoAN2JxEqdGPxFMAQSzAA9lgwNtzCQGYArLO4pFygGbHjwgBQxyNYgIBKBAA+EXQAKRIAOnljGk8MAAlKDCgAajTYXwBtHQIAXQC5FDdjT08VKEJQwICgA

```
type a = {
  "name": string,
  "age": int
}

type b = {
  "name": string 
}

let v: a = {
  "name": "Anton",
  "age": 35 
}

let foo = (arg: b) => {
  Js.log("Hi" ++ arg["name"])
}

foo((v :> b))
```